File: 127-word-ladder/word-ladder.java
class Pair {
    String word;
    int length;

    public Pair(String word, int length) {
        this.word = word;
        this.length = length;
    }
}

class Solution {
    public int ladderLength(String beginWord, String endWord, List<String> wordList) {
        Set<String> set = new HashSet<>(wordList);

        Queue<Pair> q = new LinkedList<>();
        q.offer(new Pair(beginWord, 1));

        while (!q.isEmpty()) {
            Pair p = q.poll();
            String word = p.word;
            int length = p.length;

            if (Objects.equals(word , endWord))
                return length;

            for (int i = 0; i < word.length(); i++) {
                StringBuilder newWord = new StringBuilder(word);
                for (char ch = 'a'; ch <= 'z'; ch++) {
                    newWord.setCharAt(i, ch);

                    if (set.contains(newWord.toString())) {
                        q.offer(new Pair(newWord.toString(), length + 1));
                        set.remove(newWord.toString());
                    }
                }
            }
        }

        return 0;
    }
}

File: 127-word-ladder/word-ladder.py
class Solution:
    def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:

        q = deque()
        q.append((beginWord, 1))
        vis = set()
        words = set(wordList)

        while q:
            word, n = q.popleft()

            if word == endWord:
                return n
            
            word_arr = list(word)

            for i,letter in enumerate(word_arr):
                for c in range(ord('a') , ord('z') + 1):
                    word_arr[i] = chr(c)

                    new_word = "".join(word_arr)
                    if new_word in words and new_word not in vis:
                        vis.add(new_word)
                        q.append((new_word, n + 1))
                
                word_arr[i] = letter

        return 0
        